File: apps/backend/app/schemas/portfolio.py
from pydantic import BaseModel, Field, field_validator


class HoldingInput(BaseModel):
    currency_code: str = Field(min_length=3, max_length=3)
    weight_percent: float = Field(gt=0, le=100)

    @field_validator("currency_code")
    @classmethod
    def uppercase_currency(cls, value: str) -> str:
        return value.upper()


class CreatePortfolioRequest(BaseModel):
    base_currency: str = Field(default="USD", min_length=3, max_length=3)

    @field_validator("base_currency")
    @classmethod
    def uppercase_base_currency(cls, value: str) -> str:
        return value.upper()


class UpdateBaseCurrencyRequest(BaseModel):
    base_currency: str = Field(min_length=3, max_length=3)

    @field_validator("base_currency")
    @classmethod
    def uppercase_base_currency(cls, value: str) -> str:
        return value.upper()


class UpdateHoldingsRequest(BaseModel):
    holdings: list[HoldingInput] = Field(min_length=1)

    @field_validator("holdings")
    @classmethod
    def weights_sum_to_one_hundred(cls, holdings: list[HoldingInput]) -> list[HoldingInput]:
        total = round(sum(item.weight_percent for item in holdings), 4)
        if total != 100.0:
            raise ValueError("holdings weights must sum to 100")
        return holdings


class HoldingResponse(BaseModel):
    currency_code: str
    weight_percent: float
    quantity: float


class HoldingDetailResponse(HoldingResponse):
    usd_value: float
    weight_actual_percent: float


class HoldingDeltaResponse(BaseModel):
    currency_code: str
    quantity_delta: float
    usd_delta: float


class PortfolioResponse(BaseModel):
    id: str
    base_currency: str
    initial_cash_usd: float
    total_value_usd: float
    daily_pl_usd: float
    rates_date: str | None
    prior_rates_date: str | None
    holdings: list[HoldingResponse]
    holdings_detail: list[HoldingDetailResponse]


class PreviewHoldingsResponse(BaseModel):
    base_currency: str
    total_value_usd: float
    projected_holdings: list[HoldingDetailResponse]
    deltas: list[HoldingDeltaResponse]


class PortfolioHistoryPointResponse(BaseModel):
    date: str
    total_value_usd: float
    daily_pl_usd: float


class PortfolioHistoryResponse(BaseModel):
    base_currency: str
    points: list[PortfolioHistoryPointResponse]
    rebalance_markers: list[str]


class PortfolioSnapshotResponse(BaseModel):
    base_currency: str
    as_of: str | None
    total_value_usd: float
    daily_pl_usd: float
    holdings: list[HoldingDetailResponse]
    disclaimer: str


class PortfolioTransactionHoldingResponse(BaseModel):
    currency_code: str
    weight_percent: float
    quantity: float


class PortfolioTransactionResponse(BaseModel):
    id: int
    event_type: str
    base_currency: str
    effective_rates_date: str
    total_value_usd: float
    holdings: list[PortfolioTransactionHoldingResponse]
    created_at: str


class PortfolioTransactionsResponse(BaseModel):
    base_currency: str
    transactions: list[PortfolioTransactionResponse]

File: apps/frontend/lib/api/portfolio.ts
import { apiFetch } from "./client";

export type Holding = {
  currency_code: string;
  weight_percent: number;
  quantity: number;
};

export type HoldingDetail = Holding & {
  usd_value: number;
  weight_actual_percent: number;
};

export type HoldingDelta = {
  currency_code: string;
  quantity_delta: number;
  usd_delta: number;
};

export type Portfolio = {
  id: string;
  base_currency: string;
  initial_cash_usd: number;
  total_value_usd: number;
  daily_pl_usd: number;
  rates_date: string | null;
  prior_rates_date: string | null;
  holdings: Holding[];
  holdings_detail: HoldingDetail[];
};

export type AllocationInput = {
  currency_code: string;
  weight_percent: number;
};

export type PreviewHoldingsResponse = {
  base_currency: string;
  total_value_usd: number;
  projected_holdings: HoldingDetail[];
  deltas: HoldingDelta[];
};

export type PortfolioHistoryPoint = {
  date: string;
  total_value_usd: number;
  daily_pl_usd: number;
};

export type PortfolioHistoryResponse = {
  base_currency: string;
  points: PortfolioHistoryPoint[];
  rebalance_markers: string[];
};

export type PortfolioSnapshotResponse = {
  base_currency: string;
  as_of: string | null;
  total_value_usd: number;
  daily_pl_usd: number;
  holdings: HoldingDetail[];
  disclaimer: string;
};

export type PortfolioTransactionHolding = {
  currency_code: string;
  weight_percent: number;
  quantity: number;
};

export type PortfolioTransaction = {
  id: number;
  event_type: "initial" | "rebalance";
  base_currency: string;
  effective_rates_date: string;
  total_value_usd: number;
  holdings: PortfolioTransactionHolding[];
  created_at: string;
};

export type PortfolioTransactionsResponse = {
  base_currency: string;
  transactions: PortfolioTransaction[];
};

const PORTFOLIO_STORAGE_KEY = "marketmage-portfolio-id";

export function getStoredPortfolioId(): string | null {
  if (typeof window === "undefined") {
    return null;
  }
  return window.localStorage.getItem(PORTFOLIO_STORAGE_KEY);
}

export function storePortfolioId(id: string): void {
  window.localStorage.setItem(PORTFOLIO_STORAGE_KEY, id);
}

export async function createPortfolio(baseCurrency = "USD"): Promise<Portfolio> {
  return apiFetch<Portfolio>("/v1/portfolio", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ base_currency: baseCurrency }),
  });
}

export async function fetchPortfolio(portfolioId: string): Promise<Portfolio> {
  return apiFetch<Portfolio>(`/v1/portfolio/${portfolioId}`);
}

export async function updatePortfolioHoldings(
  portfolioId: string,
  holdings: AllocationInput[],
): Promise<Portfolio> {
  return apiFetch<Portfolio>(`/v1/portfolio/${portfolioId}/holdings`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ holdings }),
  });
}

export async function previewPortfolioHoldings(
  portfolioId: string,
  holdings: AllocationInput[],
): Promise<PreviewHoldingsResponse> {
  return apiFetch<PreviewHoldingsResponse>(`/v1/portfolio/${portfolioId}/holdings/preview`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ holdings }),
  });
}

export async function fetchPortfolioHistory(
  portfolioId: string,
  days = 30,
): Promise<PortfolioHistoryResponse> {
  return apiFetch<PortfolioHistoryResponse>(`/v1/portfolio/${portfolioId}/history?days=${days}`);
}

export async function fetchPortfolioSnapshot(portfolioId: string): Promise<PortfolioSnapshotResponse> {
  return apiFetch<PortfolioSnapshotResponse>(`/v1/portfolio/${portfolioId}/snapshot`);
}

export async function fetchPortfolioTransactions(
  portfolioId: string,
): Promise<PortfolioTransactionsResponse> {
  return apiFetch<PortfolioTransactionsResponse>(`/v1/portfolio/${portfolioId}/transactions`);
}

export async function switchPortfolioBaseCurrency(
  portfolioId: string,
  baseCurrency: string,
): Promise<Portfolio> {
  return apiFetch<Portfolio>(`/v1/portfolio/${portfolioId}/base-currency`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ base_currency: baseCurrency }),
  });
}

export function formatUsd(value: number): string {
  return formatMoney(value, "USD");
}

export function formatMoney(value: number, currency: string): string {
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency,
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(value);
}

export const ALLOCATABLE_CURRENCIES = ["USD", "EUR", "JPY", "GBP", "CNY", "SGD"] as const;
export const BASE_CURRENCY_OPTIONS = ALLOCATABLE_CURRENCIES;
